openmls/src/extensions/mod.rs

```diff
@@ -195,6 +195,183 @@ pub enum Extension {
+
+    // ---------------------------------------------------------------------------------------------
+
+    /// Add an extension to the extension list.
+    ///
+    /// Returns an error when there already is an extension with the same extension type.
+    pub fn add(&mut self, extension: Extension) -> Result<(), &'static str> {
+        if !self.contains(extension.extension_type()) {
+            self.inner.push(extension);
+            Ok(())
+        } else {
+            Err("Cannot add duplicate extension.")
+        }
+    }
+
+    /// Add an extension to the extension list (or silently replace an existing one.)
+    pub fn add_or_replace(&mut self, extension: Extension) {
+        let _ = self.remove(extension.extension_type());
+        self.add(extension).unwrap();
+    }
+
+    /// Remove an extension from the extension list.
+    ///
+    /// Returns an error when there is no extension with the given extension type.
+    pub fn remove(&mut self, extension_type: ExtensionType) -> Result<(), &'static str> {
+        if self.contains(extension_type) {
+            self.inner.retain(|e| e.extension_type() != extension_type);
+            Ok(())
+        } else {
+            Err("Cannot remove non-existent extension.")
+        }
+    }
+
+    /// Replace an extension in the extension list.
+    ///
+    /// Returns an error when there is no extension with the given extension type.
+    #[cfg(any(feature = "test-utils", test))]
+    pub fn replace(&mut self, extension: Extension) -> Result<(), &'static str> {
+        if self.contains(extension.extension_type()) {
+            self.remove(extension.extension_type()).unwrap();
+            self.add(extension).unwrap();
+            Ok(())
+        } else {
+            Err("Cannot replace non-existent extension.")
+        }
+    }
+
+    // ---------------------------------------------------------------------------------------------
+
+    /// Return true if the extension list contains the extension with the given type.
+    pub fn contains(&self, extension_type: ExtensionType) -> bool {
+        self.inner
+            .iter()
+            .any(|e| e.extension_type() == extension_type)
+    }
+
+    /// Check that the candidate extension list is valid.
+    ///
+    /// Valid means:
+    ///
+    /// * Does not contain duplicate extension types (ValSem012)
+    pub fn validate(candidate: &[Extension]) -> bool {
+        // We use a [`HashSet`] to identify duplicate extension types by ...
+        let mut hash_map = HashSet::new();
+
+        for extension_type in candidate.iter().map(Extension::extension_type) {
+            // ... trying to insert every element ...
+            if !hash_map.insert(extension_type) {
+                // ... and returning false if an element of the same extension type was inserted before.
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl TryFrom<Vec<Extension>> for Extensions {
+    type Error = &'static str;
```

Comment:
Don't use string errors. There's an error for this already you should re-use (`InvalidExtensionError`).